### PR TITLE
 GEODE-6649: Remove lock contention in ClientHealthMonitor heartbeats.

### PR DESCRIPTION
**Pull request by pivotal-jbarrett**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java [hunk 1]
```diff
@@ -351,18 +333,19 @@ public void removeConnection(ClientProxyMembershipID proxyID, ServerConnection c
...
       logger.trace("ClientHealthMonitor: Received ping from client with member id {}", proxyID);
     }
-    synchronized (_clientHeartbeatsLock) {
-      if (!this._clientHeartbeats.containsKey(proxyID)) {
-        registerClient(proxyID);
-      } else {
-        this._clientHeartbeats.put(proxyID, Long.valueOf(System.currentTimeMillis()));
-      }
+
+    AtomicLong heartbeat = _clientHeartbeats.get(proxyID);
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: It seems like this logic should all be in registerClient and this code should just call registerClient. I don't understand why the old code split this code up into two places.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java [hunk 2]
```diff
@@ -398,29 +381,21 @@ public void receivedPing(ClientProxyMembershipID proxyID) {
...
+          String clientString;
           if (socketAddress == null) {
             clientString = "client member id=" + membershipID;
           } else {
             clientString = "host name=" + socketAddress.toString() + " host ip="
                 + socketAddress.getHostAddress() + " client port=" + socketPort
                 + " client member id=" + membershipID;
           }
-          Object[] data = null;
+          Object[] data;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: unite lines 392 and 393: Object [] data - map.get(membershipID);

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java [hunk 3]
```diff
@@ -195,25 +195,18 @@ public static synchronized void shutdownInstance() {
...
-        this._clientHeartbeats = newClientHeartbeats;
-        registerClient = true;
-      }
-    }
-
-    if (registerClient) {
-      if (this.stats != null) {
-        this.stats.incClientRegisterRequests();
+    AtomicLong heartbeat =
+        _clientHeartbeats.putIfAbsent(proxyID, new AtomicLong(System.currentTimeMillis()));
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: Shoot, I meant to use `computeIfAbsent`. Good catch, thanks.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java [hunk 4]
```diff
@@ -195,25 +195,18 @@ public static synchronized void shutdownInstance() {
...
+    if (heartbeat == null) {
+      if (stats != null) {
+        stats.incClientRegisterRequests();
       }
       if (logger.isDebugEnabled()) {
         logger.debug("ClientHealthMonitor: Registering client with member id {}",
             proxyID);
       }
+    } else {
+      heartbeat.set(System.currentTimeMillis());
```
[the remark this answers is not in the document]
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: Good idea!

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java (hunk 1 above)
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: I recall there being some other code paths that came in where it made sense to split them. Something about registering without updating the time or something like that. Basically I tried to leave the logic the same and just pull out the locking. Wasn't looking to refactor the insanity in here.

Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: But this code in receivePing is basically just duplicating logic in registerClient. If it finds an entry in clientHeartbeats it updates it by setting it to currentTimeMillis. If it doesn't find one then it calls registerClient which will then just do all that again. I don't see any value in having this logic duplicated. The only difference is we call "get" here first (which is what I suggested we call first in registerClient).

Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: Ok, I see it now. When I was whittling down the old methods I didn't see this optimization. Thanks!

Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java]: Actually it turns out the original didn't have an update time in register if it was already registered. Added tests to constrain the refactor and cleaned this up a bit more.

File: geode-cq/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/DurableClientTestBase.java [hunk 5]
```diff
@@ -215,7 +215,7 @@ public void run2() throws CacheException {
                 .info(st + " CCP states: " + getAllClientProxyState());
             CacheServerTestUtil.getCache().getLogger().info(st + " CHM states: "
                 + printMap(
-                    ClientHealthMonitor._instance.getConnectedClients(null)));
+                    ClientHealthMonitor.instance.getConnectedClients(null)));
```
Comment (dschneider-pivotal) [inline geode-cq/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/DurableClientTestBase.java]: how about changing "instance" to be private and have this code use the public "getInstance()"?

**Pull request merged**